File: convert-sorted-array-to-binary-search-tree/convert-sorted-array-to-binary-search-tree.cc
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
  int val;
  TreeNode *left, right;
  TreeNode(int x): val(x), left(NULL), right(NULL) {}
};

typedef TreeNode *node;
class Solution {
public:
  node sa2bst(vector<int> &num, int s, int e) {
    if(s == e) return NULL;
    int mid = (s + e) / 2;
    node left = sa2bst(num, s, mid), right = sa2bst(num, mid+1, e);
    node root = new TreeNode(num[mid]);
    root->left = left, root->right = right;
    return root;
  }
  node sortedArrayToBST(vector<int> &num) {
    return sa2bst(num, 0, num.size());
  }
};

File: rs/src/quests/convert_sorted_array_to_binary_search_tree.rs
/**
 * [108] Convert Sorted Array to Binary Search Tree
 *
 * Given an integer array nums where the elements are sorted in ascending order, convert it to a height-balanced binary search tree.
A height-balanced binary tree is a binary tree in which the depth of the two subtrees of every node never differs by more than one.
 
Example 1:
Input: nums = [-10,-3,0,5,9]
Output: [0,-3,9,-10,null,5]
Explanation: [0,-10,5,null,-3,null,9] is also accepted:

Example 2:
Input: nums = [1,3]
Output: [3,1]
Explanation: [1,null,3] and [3,1] are both height-balanced BSTs.

 
Constraints:

	1 <= nums.length <= 104
	-104 <= nums[i] <= 104
	nums is sorted in a strictly increasing order.

 */
pub struct Solution {}
use super::super::util::tree::{TreeNode, to_tree};

// submission codes start here

// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn sorted_array_to_bst(nums: Vec<i32>) -> Option<Rc<RefCell<TreeNode>>> {
        if nums.is_empty() { return None; }

        let n = nums.len();
        let mid = n >> 1;
        let (l, r) = nums.split_at(mid);
        let left = Self::sorted_array_to_bst(l.to_vec());
        let right = Self::sorted_array_to_bst(r[1..].to_vec());
        let root = Rc::new(RefCell::new(TreeNode {
            val: nums[mid],
            left,
            right,
        }));
        Some(root)
    }
}

// submission codes end

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() {
    }
}
